db/version_builder.cc

```diff
@@ -224,52 +224,65 @@ class VersionBuilder::Rep {
     return Status::OK();
   }
 
-  Status CheckConsistencyForDeletes(VersionEdit* /*edit*/, uint64_t number,
-                                    int level) {
+  Status CheckConsistencyForDeletes(VersionEdit* edit) {
 #ifdef NDEBUG
     if (!base_vstorage_->force_consistency_checks()) {
       // Dont run consistency checks in release mode except if
       // explicitly asked to
       return Status::OK();
     }
 #endif
+    std::unordered_map<uint64_t, int> deletes;
+    for (const auto& del_file : edit->GetDeletedFiles()) {
+      const auto level = del_file.first;
+      const auto number = del_file.second;
+      if (level < num_levels_) {
+        deletes[number] = level;
+      }
+    }
+
     // a file to be deleted better exist in the previous version
-    bool found = false;
-    for (int l = 0; !found && l < num_levels_; l++) {
+    for (int l = 0; deletes.size() != 0 && l < num_levels_; l++) {
       const std::vector<FileMetaData*>& base_files =
           base_vstorage_->LevelFiles(l);
       for (size_t i = 0; i < base_files.size(); i++) {
         FileMetaData* f = base_files[i];
-        if (f->fd.GetNumber() == number) {
+        if (deletes.erase(f->fd.GetNumber()) != 0) {
+          if (deletes.size() == 0) {
+            break;
+          }
+        }
+      }
+    }
+
+    for (const auto& d : deletes) {
+      const auto number = d.first;
+      const auto level = d.second;
+      // if the file did not exist in the previous version, then it
+      // is possibly moved from lower level to higher level in current
+      // version
+      bool found = false;
+      for (int l = level + 1; l < num_levels_; l++) {
+        auto& level_added = levels_[l].added_files;
+        auto got = level_added.find(number);
+        if (got != level_added.end()) {
           found = true;
           break;
         }
       }
-    }
-    // if the file did not exist in the previous version, then it
-    // is possibly moved from lower level to higher level in current
-    // version
-    for (int l = level + 1; !found && l < num_levels_; l++) {
-      auto& level_added = levels_[l].added_files;
-      auto got = level_added.find(number);
-      if (got != level_added.end()) {
-        found = true;
-        break;
+      // maybe this file was added in a previous edit that was Applied
+      if (!found) {
+        auto& level_added = levels_[level].added_files;
```

Comment:
not sure if I misread, but can the block of 264-272 and 274-279 be merged into one? (loop l = [level .. num_levels_])

Reply:
yes, but original code writes that way which seems to make the logic more clear.

Reply:
make sense too.